From c4407898217d185d83e278ce18d5f2c40c144042 Mon Sep 17 00:00:00 2001
From: Ikey Doherty <michael.i.doherty@intel.com>
Date: Mon, 7 Mar 2016 09:38:18 +0000
Subject: [PATCH] Support an entirely stateless configuration

Signed-off-by: Ikey Doherty <michael.i.doherty@intel.com>
---
 src/up-config.c | 46 ++++++++++++++++++++++++++++++----------------
 1 file changed, 30 insertions(+), 16 deletions(-)

diff --git a/src/up-config.c b/src/up-config.c
index 2a17f30..ad7f0c1 100644
--- a/src/up-config.c
+++ b/src/up-config.c
@@ -99,28 +99,42 @@ up_config_init (UpConfig *config)
 {
 	gboolean ret;
 	GError *error = NULL;
-	gchar *filename;
+	gchar *filename = NULL;
+	gchar *paths[3] = { 0 };
 
 	config->priv = up_config_get_instance_private (config);
 	config->priv->keyfile = g_key_file_new ();
 
-	filename = g_strdup (g_getenv ("UPOWER_CONF_FILE_NAME"));
-	if (filename == NULL)
-		filename = g_build_filename (PACKAGE_SYSCONF_DIR,"UPower", "UPower.conf", NULL);
-
-	/* load */
-	ret = g_key_file_load_from_file (config->priv->keyfile,
-					 filename,
-					 G_KEY_FILE_NONE,
-					 &error);
-
-	if (!ret) {
-		g_warning ("failed to load config file '%s': %s",
-			   filename, error->message);
-		g_error_free (error);
+	paths[0] = g_strdup (g_getenv ("UPOWER_CONF_FILE_NAME"));
+	paths[1] = g_build_filename (PACKAGE_SYSCONF_DIR, "UPower", "UPower.conf", NULL);
+	paths[2] = g_build_filename ("/usr/share/upower/", "UPower", "UPower.conf", NULL);
+
+	for (int i = 0; i < 3; i++) {
+		filename = paths[i];
+		if (!filename) {
+			continue;
+		}
+		ret = g_key_file_load_from_file (config->priv->keyfile,
+						 filename,
+						 G_KEY_FILE_NONE,
+						 &error);
+		if (!ret) {
+			if (i == 2) {
+				g_warning("failed to load config file '%s': %s",
+					filename, error->message);
+			}
+			g_error_free(error);
+		}
+		/* Found a valid configuration */
+		break;
 	}
 
-	g_free (filename);
+	/* Free paths */
+	for (int i = 0; i < 3; i++) {
+		if (paths[i]) {
+			g_free(paths[i]);
+		}
+	}
 }
 
 /**
-- 
2.7.1

